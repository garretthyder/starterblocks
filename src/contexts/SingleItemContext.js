import { createContext } from '@wordpress/element';
const SingleItemContext = createContext({});
export const SingleItemProvider = SingleItemContext.Provider;
export const SingleItemConsumer = SingleItemContext.Consumer;
export default SingleItemContext;
