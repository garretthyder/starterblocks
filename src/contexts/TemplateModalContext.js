import { createContext } from '@wordpress/element';
const TemplateModalContext = createContext({});
export const TemplateModalProvider = TemplateModalContext.Provider;
export const TemplateModalConsumer = TemplateModalContext.Consumer;
export default TemplateModalContext;
