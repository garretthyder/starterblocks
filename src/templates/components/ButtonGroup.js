const {__} = wp.i18n
const {Component, Fragment, useContext, useEffect, useState} = wp.element;
import SingleItemContext from '../contexts/SingleItemContext';
import TemplateModalContext from '../contexts/TemplateModalContext';
import {missingPro, missingRequirement} from '../stores/helper';

export default function ButtonGroup(props) {
    const {data, activeItemType, index, pageData} = useContext(SingleItemContext);
    const {openSitePreviewModal, onImportTemplate, spinner} = useContext(TemplateModalContext);
    const [rootClassName, setRootClassName] = useState('starterblocks-import-button-group');
    
    const {ID, image, url, pro, requirements} = data;

    useEffect(() => {
        if (spinner === null && rootClassName !== 'starterblocks-import-button-group')
            setRootClassName('starterblocks-import-button-group')
        if (spinner !== null && rootClassName === 'starterblocks-import-button-group')
            setRootClassName('starterblocks-import-button-group disabled');

    }, [spinner])

    const triggerImportTemplate = (data) => {
        if (spinner === null) onImportTemplate(data);
    }

    const isMissingRequirement = missingRequirement(pro, requirements);
    const isMissingPro = missingPro(pro);

    return (
        <div className={rootClassName}>
            <Fragment>
                {url &&
                <a className="starterblocks-button" target="_blank"
                    onClick={() => openSitePreviewModal(index, pageData)}>
                    <i className="fa fa-share"/> {__('Preview')}
                </a>
                }
                <a className="starterblocks-button starterblocks-button-download"
                    onClick={() => triggerImportTemplate(data)}>
                    {spinner === ID ? <i className="fas fa-spinner fa-pulse"/> :
                        <i className="fas fa-download"/>}{__('Import')}
                </a>
            </Fragment>
        </div>
    )
}
