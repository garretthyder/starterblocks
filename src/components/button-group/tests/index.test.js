import { shallow, mount } from 'enzyme';
import { render } from '@testing-library/react'

import ButtonGroup from '../';
import {TemplateModalProvider} from '../../../contexts/TemplateModalContext';
import ShallowRenderer from 'react-test-renderer/shallow';
import renderer from 'react-test-renderer';

// To be used fro snapshot
const TestComponent = (props) => {
    const {spinner} = props;
    return (
        <TemplateModalProvider value={{spinner}}>
            <ButtonGroup/>
        </TemplateModalProvider>
    );
}

jest.mock('../../preview-import', () => () => 'PreviewImport');
jest.mock('../../dependent-plugins', () => () => 'DependentPlugins');


let realUseContext, useContextMock;
let useEffect;

describe('Button Group', () => {
    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    };

    beforeEach(() => {
        useEffect = jest.spyOn(React, 'useEffect');
    });

    afterEach(() => {

    });

    // 1. Snapshot testing
    it('renders correctly', () => {
        // const element = shallow(<TestComponent spinner={null} />);
        const component = renderer.create(<TestComponent spinner={null} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        // expect(element).toMatchSnapshot();
    });

    // 2. Testing Props: default Prop
    it('renders the default classname with spinner null', () => {
        realUseContext = React.useContext;
        useContextMock = React.useContext = jest.fn();
        useContextMock.mockReturnValue({spinner: null});
        const renderer = new ShallowRenderer();
        const element = renderer.render(<ButtonGroup />);
        expect(element.props.className).toBe('starterblocks-import-button-group');
        React.useContext = realUseContext;
    })

    it('renders disabled status with spinner not null', () => {
        realUseContext = React.useContext;
        useContextMock = React.useContext = jest.fn();
        mockUseEffect();
        useContextMock.mockReturnValue({spinner: 1});
        const element = new ShallowRenderer().render(<ButtonGroup />);
        expect(element.props.className).toBe('starterblocks-import-button-group disabled');
        React.useContext = realUseContext;
    })

    // 3. Data types:

});
