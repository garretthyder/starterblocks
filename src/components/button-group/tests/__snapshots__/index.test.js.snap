// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Group renders correctly 1`] = `
<div
  className="starterblocks-import-button-group"
>
  PreviewImport
  DependentPlugins
</div>
`;
